src/runtime_src/core/common/api/xrt_module.cpp

```diff
@@ -710,6 +781,21 @@ class module_elf : public module_impl
     return true;
   }
 
+public:
+  bool
+  patch_it(uint8_t* base, const std::string& argnm, size_t index, uint64_t patch,
+           patcher::buf_type type, uint32_t sec_index) override
+  {
+    return patch_it_impl(base, argnm, index, patch, type, sec_index);
+  }
+
+  bool
+  patch_it(xrt::bo bo, const std::string& argnm, size_t index, uint64_t patch,
+           patcher::buf_type type, uint32_t sec_index) override
+  {
+    return patch_it_impl(bo, argnm, index, patch, type, sec_index);
+  }
```

Comment:
```suggestion
  bool
  patch_it(const xrt::bo& bo, const std::string& argnm, size_t index, uint64_t patch,
           patcher::buf_type type, uint32_t sec_index) override
  {
    return patch_it_impl(bo.map<uint8_t*>(), argnm, index, patch, type, sec_index);
  }
```

Reply:
Hi @stsoe, we need to call sync on the bo for the args that are patched.
My initial implementation was to have 2 separate functions so that I can do bo specific code but refactored it this way to avoid code duplication. I can have 2 functions if code duplication is okay.

The other way to solve this is :

bool
patch_it(xrt::bo bo, ....)
{
  if (!patch_it_impl(bo.map<>....))
     return false;

  for (...) {
     switch (m_symbol_type) {
        case a :
          bo.sync();    // call sync here
        case b :
           bo.sync();
        ....
     }
  }
}

Let me know your thoughts, Thanks

Reply:
I see.  The template just isn't right in this case, but neither is code duplication.  Two options (1) different kind of refactoring, (2) std::variant with std::visit for the special action depending on the variant type.

Reply:
Thanks Soren for the suggestion, refactored the code to use std::variant and std::visit.
Please review